lib/puppet/provider/openldap_schema/olc.rb

```diff
@@ -0,0 +1,123 @@
+require 'tempfile'
+
+Puppet::Type.type(:openldap_schema).provide(:olc) do
+
+  # TODO: Use ruby bindings (can't find one that support IPC)
+
+  defaultfor :osfamily => :debian, :osfamily => :redhat
+
+  commands :slapcat => 'slapcat', :ldapadd => 'ldapadd', :slaptest => 'slaptest', :ldapsearch => 'ldapsearch'
+
+  mk_resource_methods
+
+  @@openldap_schema_existing = []
+  @@openldap_schema_new = []
+  @@openldap_schema_create_count = 0
+  @@openldap_schema_max_count = 0
+
+  def self.instances
+    self.getSchemaNames.map { |schema|
+          new(
+            :ensure => :present,
+            :name   => schema
+          )
+    }
+  end
+
+  def self.getSchemaNames
+      schemas = []
+      ldapsearch('-Y', 'EXTERNAL', '-H', 'ldapi:///', '-b', 'cn=config', '(objectClass=olcSchemaConfig)', 'cn').split("\n\n").each do |paragraph|
+          paragraph.split("\n").each do |line|
+              if line =~ /^cn: \{/
+                  schemas.push line
+              end
+          end
+      end
+      schemas.map{ |entry| entry.match(/^cn: \{\d+\}(\S+)/)[1] }
+  end
+
+  def self.prefetch(resources)
+    # Match installed schemas with requested schemas to find the paths to them
+    installed = getSchemaNames()
+    for schema in installed do
+        if match = resources.values.find{ |resource| resource.pathWithDefault =~ /#{Regexp.escape(schema)}\.schema$/ }
+            @@openldap_schema_existing.push(match)
+        else
+            warning("Missing schema '#{schema}', might cause insertion issues")
+        end
+    end
+
+    mods = instances
+    
+    resources.keys.each do |name|
+      if provider = mods.find{ |mod| mod.name == name }
+        resources[name].provider = provider
+      else
+        @@openldap_schema_max_count += 1
+      end
+    end
+  end
+
+  def create
+    # Order new resources by priority
+    @@openldap_schema_new.push(resource)
+
+    # Only insert in the last requested resource
+    @@openldap_schema_create_count += 1
+    puts @@openldap_schema_max_count
+    unless @@openldap_schema_create_count == @@openldap_schema_max_count
+        return
+    end
+
+    t = Tempfile.new('openldap_schemas_ldif')
+    t2 = Tempfile.new('openldap_schemas_includes')
+    
+    for resource in @@openldap_schema_existing do
+        t2 << "include #{resource.pathWithDefault}\n"
```

Comment:
Can't you use `slaptest` to convert schema to a temporary ldif file and then just can `ldapadd` on it? I'm looking for a way to avoid the class variables...